spec/eofv0_verkle.md

```diff
@@ -0,0 +1,107 @@
+# EOFv0 for packaging legacy code in Verkle Trees
+
+The design draft that proposes the use of EOF 
+for storing code in Verkle Trees.
+An alternative to the existing method of executing 
+31-byte code chunks accompanied by 1 byte of metadata.
+
+## Goal
+
+Provide the result of the jumpdest analysis of a deployed code as the EOF section.
+During code execution the jumpdest analysis is already available
+and the answer to the question "is this jump target valid?" can be looked up
+in the section. This allows using 32-byte Verkle Tree code chunks
+(instead of 31-byte of code + 1 byte of metadata).
+
+## Specification Draft
+
+1. Put the code in the single *code* EOF section.
+2. Use the EOF container format proposed by [EIP-3540](https://eips.ethereum.org/EIPS/eip-3540) with
+   version 0 and following modifications to "Changes to execution semantics":
+    1. `CODECOPY`/`CODESIZE`/`EXTCODECOPY`/`EXTCODESIZE`/`EXTCODEHASH` operates on the *code*
+       section only.
+    2. `JUMP`/`JUMPI`/`PC` relates code positions to the *code* section only.
+3. Perform the jumpdest analysis of the code at deploy time (during contract creation).
```

Comment:
This actually makes a big semantical change here: we are locking in the EVM version at the time of contract deployment (or verkle transition).

Currently in mainnet we rely on the fact that semantics of contracts can change. This is both a negative (and maybe a positive?)

Example: if a new opcode is introduced, the jumpdest analysis result of a contract may change. This is not the case after this proposal.

Reply:
I've not been following 100%, but could we mitigate this:

> Example: if a new opcode is introduced, the jumpdest analysis result of a contract may change. This is not the case after this proposal.

by versioning the jumpdest analysis result and updating it on first use after the new opcode was introduced? Similar to how we do "packaging" (point 5), we do "re-packaging".

Reply:
Not sure what you mean -- I mean that if the jumpdest analysis result is locked it at the time of contract creation / verkle transition, then introduction of new legacy opcodes will need to have special considerations for such contracts.

Currently we can do an analysis on chain to see what effect a new opcode may be (does it change semantics of contracts?), but with the addition of the jumpdest table this analysis is different.

Reply:
I mean that when you store the jumpdest analysis result during verkle transition, you can prepend a version `1` to it. If there's a new opcode introduced you would need to re-do the analysis, overwrite the jumpdest result and bump the version. That would happen the next time the code is touched.

Reply:
That is unfeasible.  Would need to update the entire change at every hardfork.  This is one of the reasons any transition like verkle or flat-tree proposed earlier hits the roadblock.

Reply:
Interesting topic. Locking jumpdest analysis allows introducing instructions with immediate data. But this only works for deployed code but not for initcode (where analysis runs before execution).